crucible-symio/src/Lang/Crucible/SymIO/Types.hs

```diff
@@ -0,0 +1,208 @@
+  . MapF.insert (knownSymbol :: SymbolRepr "VFS_file") IntrinsicMuxFn
+  . MapF.insert (knownSymbol :: SymbolRepr "VFS_filepointer") IntrinsicMuxFn
+  $ MapF.empty
+
+-- | An identifier for a file, which must be resolved into a 'File' to access
+-- the underlying filesystem.
+--
+-- This is a file path
+type FileIdent sym = RegValue sym FileIdentType
+
+-- | The crucible-level type of 'FileIdent'
+type FileIdentType = StringType Char8
+
+-- | The crucible-level type of 'FileSystem'
+type FileSystemType w = IntrinsicType "VFS_filesystem" (EmptyCtx ::> BVType w)
+
+-- | Defines the current state of a symbolic filesystem.
+data FileSystem sym w =
+  FileSystem
+    {
+      fsPtrSize :: NatRepr w
+    , fsFileNames :: RegValue sym (StringMapType (FileType w))
+    -- ^ map from concrete file identifiers to files
+    , fsFileSizes :: CA.CachedArray sym (EmptyCtx ::> BaseIntegerType) (BaseBVType w)
+    -- ^ a symbolic map from files to their size
+    , fsSymData :: CA.CachedArray sym (EmptyCtx ::> BaseBVType w ::> BaseIntegerType) (BaseBVType 8)
+    -- ^ array representing symbolic file contents
+    , fsConstraints :: forall a. ((IsSymInterface sym, 1 <= w) => a) -> a
+    }
+
+-- | A base index into the filesystem, consistent of a file identifier and an offset into that file.
+type FileSystemIndex sym w = Assignment (SymExpr sym) (EmptyCtx ::> BaseBVType w ::> BaseIntegerType)
+
+muxFileSystem ::
+  IsSymInterface sym =>
+  sym ->
+  Pred sym ->
+  FileSystem sym w ->
+  FileSystem sym w ->
+  IO (FileSystem sym w)
+muxFileSystem sym p fsT fsF = do
+  symData <- CA.muxArrays sym p (fsSymData fsT) (fsSymData fsF)
+  symFiles <- muxStringMap sym (muxFile sym) p (fsFileNames fsT) (fsFileNames fsF)
+  symFileSizes <- CA.muxArrays sym p (fsFileSizes fsT) (fsFileSizes fsF)
+  return $ fsT { fsSymData  = symData, fsFileNames = symFiles, fsFileSizes = symFileSizes }
+
+instance (IsSymInterface sym) => IntrinsicClass sym "VFS_filesystem" where
+  type Intrinsic sym "VFS_filesystem" (EmptyCtx ::> BVType w) = FileSystem sym w
+
+  muxIntrinsic sym _iTypes _nm (Empty :> (BVRepr _w)) = muxFileSystem sym
+  muxIntrinsic _ _ nm ctx = \_ _ _ -> typeError nm ctx
+
+pattern FileSystemRepr :: () => (1 <= w, ty ~ FileSystemType w) => NatRepr w -> TypeRepr ty
+pattern FileSystemRepr w <- IntrinsicRepr (testEquality (knownSymbol :: SymbolRepr "VFS_filesystem") -> Just Refl)
+                                           (Empty :> BVRepr w)
+  where
+    FileSystemRepr w = IntrinsicRepr knownSymbol (Empty :> BVRepr w)
+
+-- | The crucible type of file handles.
+type FileHandleType w = ReferenceType (MaybeType (FilePointerType w))
+
+-- |  A file handle is a mutable file pointer that increments every time it is read.
+type FileHandle sym w = RegValue sym (FileHandleType w)
+
+-- | A 'File' represents a file in the filesystem independent
+-- of any open handles to it
+data File sym w = File (NatRepr w) (SymInteger sym)
```

Comment:
These correspond basically to `inode`s right?

Reply:
Yes, I believe that is a good characterization